File: scala/49.scala
object Solution {
    def groupAnagrams(strs: Array[String]): List[List[String]] = {
        strs.foldLeft(Map[String, List[String]]()){
            case (res, str) => {
                val sorted = str.sorted
                if (res.contains(sorted)) {
                    res + (sorted -> (res(sorted) :+ str))
                } else {
                    res + (sorted -> List(str))
                }
            }
        }.values.toList
    }
}

File: leetcodeJava/49.java
package leetcodeJava;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        var arrayList = new ArrayList<List<String>>(Stream.of(strs).map(s -> {
            var arr = s.toCharArray();
            Arrays.sort(arr);
            return new String[]{String.valueOf(arr), s};
        }).sorted((x, y) -> {
            return x[0].compareTo(y[0]);
        }).collect(Collectors.groupingBy(x -> x[0], Collectors.mapping(x -> x[1], Collectors.toList()))).values());
        return arrayList;
    }

    public static void main(String[] args) {
        var s = new Solution();
        System.out.println(s.groupAnagrams(new String[]{"eat","tea","tan","ate","nat","bat"}));
    }
}